Some JEI, CT and Veins improvements

DisasterMoo opened the pull request:
Fixed JEI inventory tab buttons overlap (Closes #646).
Added JEI and CT integration to Chisel recipes (Closes #738).
Made category UIDs public for addons that wants to add custom recipe wrappers to TFC categories.
(ie: https://github.com/TFC-Metallum/TFCTech/blob/b33a87a180374a98d1b515d1dde0e1d0f0b4f444/src/main/java/tfctech/compat/jei/TechJEIPlugin.java#L147 where i add a custom block and pane mold for glass recipes, and hook then in the casting category)
Added JEI integration to rock layers and veins (Closes #868).
Added loose rock configuration to vein json loading (Closes #857, Closes #608).
Tweaked ore veins to be more forgiving (Closes #867).

DisasterMoo committed:
Fixed TerraFirmaCraft#646.

DisasterMoo committed:
Added CT and JEI integration to Chisel Recipes (Closes #738).
Also a little cleanup in JEI metal registration stuff for performance gains + sort by tier (this affects unmold and casting recipes, anvil catalysts already were sorted before this cleanup).

DisasterMoo committed:
Added JEI integration to rock and veins generation (Close #868).

DisasterMoo committed:
Added check to make sure vein registry won't register "air" ore veins.

DisasterMoo committed:
A few features:
- Fixed uncommon vein rarity lang entry.
- Added depth config to loose rock vein scan.
- Added loose rocks customization to vein json parsing (Closes #857).
- Updated default ore spawn data to match.

DisasterMoo committed:
Added missing "/veininfo" command. Improved "/findveins" command, now it is possible to dump all generated veins in radius + dump rate to measure how much a vein type generated.

DisasterMoo committed:
General performance improvement to find veins command.

DisasterMoo committed:
Moved log writing code to a new class for cleanup.

DisasterMoo committed:
Another improvement over find veins command: It now uses the built-in forge worker system (same system used for chunk pregen), which is called per chunk and takes into account TPS. Also, tweaked ore spawn data to be more rarity balanced (Closes #867).

DisasterMoo committed:
Added vein names (with lang entries) to vein category.

DisasterMoo committed:
Performance improvement in leaf decay algorithm.

DisasterMoo committed:
Some missing vein registration checks:
- Rarity (must be positive non-zero).
- Shape (not crash the game, just skip if not found).
Also, added some debug info output in log if config is enabled (Closes #882):
- Log about registered veins when loading.
- Log about failing at generating veins.

alcatrazEscapee commented on src/main/java/net/dries007/tfc/command/CommandFindVeins.java:
```diff
@@ -59,30 +46,165 @@ public String getName()
+        final String filter = args[0];
+
+        // Default print to player chat
+        Consumer<Chunk> consumer = chunk ->
```
Using consumers here is super ugly. Is there a reason it was done this way? I see a couple possible options that are much nicer from a code smell point of view:
 - pass these directly as arguments to the worker rather than tossing them around as variables.
 - implement these as abstract class members on the worker, then instantiate with anonymous class impl. (My preferred way)
 - add all as standard functions on the worker and just use flags to differentiate what to actually do.

alcatrazEscapee commented on src/main/java/net/dries007/tfc/command/CommandFindVeins.java:
```diff
@@ -59,30 +46,165 @@ public String getName()
+            ChunkDataTFC chunkData = ChunkDataTFC.get(chunk);
+            chunkData.getGeneratedVeins().stream()
+                .filter(vein -> !veinsFound.contains(vein.getPos()))
+                .filter(vein -> filter.equalsIgnoreCase("all") || vein.getType().getRegistryName().equalsIgnoreCase(filter))
```
If you're hyper-optimizing this for performance, this `all` check can be moved outside of the filter and only called once. Additionally, you could pre-build a `HashSet` of `VeinType`s that are required and check membership in there (`O(1)`). Probably better than string comparisons on every vein.

DisasterMoo replied:
Changed to using the VeinType obj directly, i bet that it's better than using string comparison

alcatrazEscapee commented on src/main/java/net/dries007/tfc/command/CommandFindVeins.java:
```diff
@@ -59,30 +46,165 @@ public String getName()
+                // Save files to MC's root folder
+                String fileName = "tfc-veins-rate.log";
+                Map<VeinType, Integer> veinRateMap = new HashMap<>();
+                Map<Rock, Integer> rockRateMap = new HashMap<>();
```
Again, if you're optimizing for performance, `TObjectIntHashMap` would be better

alcatrazEscapee commented on src/main/java/net/dries007/tfc/command/CommandFindVeins.java:
```diff
@@ -59,30 +46,165 @@ public String getName()
+                            {
+                                LogFileWriter.open(fileName);
+                            }
+                            LogFileWriter.writeLine(dump);
```
Is there a reason you're doing file I/O on every chunk? I can't imagine this is more efficient than simply building the data per chunk, storing it in a running `List` of some sort, then writing it all at once.

DisasterMoo replied:
It was to free memory as soon as i needn't the data anymore, but after all rewrites i did there i think it can now manage a simple `List<String>`

alcatrazEscapee commented on src/main/java/net/dries007/tfc/command/CommandFindVeins.java:
```diff
@@ -59,30 +46,165 @@ public String getName()
+            }
+            else
+            {
+                throw new WrongUsageException("3rd argument must be dump or rate");
```
Translation key

alcatrazEscapee commented on src/main/java/net/dries007/tfc/command/CommandFindVeins.java:
```diff
@@ -59,30 +46,165 @@ public String getName()
     public String getUsage(ICommandSender sender)
     {
-        return "/findveins [all|<vein name>] <radius> -> Finds all instances of a specific vein, or all veins within a certain chunk radius";
+        return "/findveins [all|<vein name>] <radius> [dump|rate] -> Finds all instances of a specific vein, or all veins within a certain chunk radius, if dump or rate save it to a log file";
```
Translation key

alcatrazEscapee commented on src/main/java/net/dries007/tfc/command/CommandFindVeins.java:
```diff
@@ -59,30 +46,165 @@ public String getName()
     public void execute(MinecraftServer server, ICommandSender sender, String[] args) throws CommandException
     {
-        if (args.length != 2) throw new WrongUsageException("2 arguments required.");
+        if (args.length != 2 && args.length != 3) throw new WrongUsageException("2 or 3 arguments required.");
```
Translation key, and be more descriptive. (i.e. see `CommandTimeTFC`)

alcatrazEscapee commented on src/main/java/net/dries007/tfc/command/CommandFindVeins.java:
```diff
@@ -99,7 +221,110 @@ public int getRequiredPermissionLevel()
+
+    private static class Worker implements WorldWorkerManager.IWorker
+    {
+        private final int dimension;
```
TFC will only ever generate veins in the overworld. This is only ever zero (including in this implementation). Any reason? Otherwise this can be removed.

DisasterMoo replied:
Sure thing. (was just a copy-paste from forge, and since i wasn't too sure if we would ever generate veins in other dimensions past 1.15, I left there)

alcatrazEscapee commented on src/main/java/net/dries007/tfc/command/CommandVeinInfo.java:
```diff
@@ -0,0 +1,60 @@
+    @Nonnull
+    public String getUsage(ICommandSender sender)
+    {
+        return "/veininfo -> Show a list of all registered veins";
```
Translation key

alcatrazEscapee commented on src/main/java/net/dries007/tfc/command/CommandVeinInfo.java:
```diff
@@ -0,0 +1,60 @@
+import net.dries007.tfc.world.classic.worldgen.vein.VeinRegistry;
+
+@ParametersAreNonnullByDefault
+public class CommandVeinInfo extends CommandBase
```
Is there a reason this is a command? All it does is log standard data. It could be quite easily just emitted at load time to the log - that would avoid having to load a new world, get in-game, run the command to check vein registration. It's a cleaner output method (i.e. already in a text file), and available as soon as veins are loaded.

DisasterMoo replied:
added it since it is referenced by `findveins` command, but i guess we can also implement it there instead

alcatrazEscapee replied:
Ah, then it should be removed there. That was likely leftover copy-paste when I first wrote that.

alcatrazEscapee commented on src/main/java/net/dries007/tfc/compat/jei/TFCJEIPlugin.java:
```diff
@@ -293,5 +335,46 @@ public void register(IModRegistry registry)
+        registry.addAdvancedGuiHandlers(new TFCGuiAreas<>(GuiSkills.class));
+    }
+
+    private static class TFCGuiAreas<T extends GuiContainer> implements IAdvancedGuiHandler<T>
```
This doesn't make any sense to be an inner class, nor does the name imo. Suggested: `TFCInventoryGuiHandler` (matches the name of the container)

alcatrazEscapee commented on src/main/java/net/dries007/tfc/compat/jei/wrappers/ChiselRecipeWrapper.java:
```diff
@@ -0,0 +1,53 @@
+                }
+            });
+        // Ideally we should use Block#getPickBlock but we can't have a World and EntityPlayer at this point
+        output = new ItemStack(recipe.getOutputState().getBlock());
```
If this fails this will be empty, could we display something separate instead? i.e. debug block maybe?

alcatrazEscapee commented on src/main/java/net/dries007/tfc/util/LogFileWriter.java:
```diff
@@ -0,0 +1,157 @@
+
+/**
+ * Handles writing text to a log file
+ * Makes code clean :D
```
I disagree strongly - this seems way more code than is nessecary here. Writing a single bunch of stuff to a file, if buffered properly:
```java
List<String> lines;
try (BufferedWriter writer = new BufferedWriter(new FileWriter(fileName))) {
    for (String s : lines) {
        writer.write(s);
    }
} catch (IOException e) {
    // log or whatever
}
```

alcatrazEscapee commented on src/main/java/net/dries007/tfc/util/json/VeinTypeJson.java:
```diff
@@ -39,11 +43,25 @@ public VeinType deserialize(JsonElement json, Type typeOfT, JsonDeserializationC
+        }
         int density = JsonUtils.getInt(jsonObject, "density");
         int width = JsonUtils.getInt(jsonObject, "width");
         int height = JsonUtils.getInt(jsonObject, "height");
```
Can add checks for nonnegative values for all these

alcatrazEscapee commented on src/main/java/net/dries007/tfc/world/classic/worldgen/vein/VeinRegistry.java:
```diff
@@ -139,8 +143,21 @@ public void postInit()
+                            TerraFirmaCraft.getLog().error("Error: " + vein.getRegistryName() + " vein registered ore block is Air! Skipping.");
+                            continue;
+                        }
+
```
There's no need for this. If the block failed to resolve to a name, it should be already caught and the vein not registered. This is only preventing the case the user wrote `"minecraft:air"` as the block, in which case why would we deny that?

DisasterMoo replied:
Not true. I found in my tests that if you specify something invalid (ie: `tfc:native_copper2`) the vein is parsed anyway containing air blocks. I can fix that in `VeinTypeJson` instead, but that will still prevent air blocks being registered: 
https://github.com/DisasterMoo/TerraFirmaCraft/blob/ef7fb80e6e89f5c618d761e637c5712b78170d73/src/main/java/net/dries007/tfc/util/json/VeinTypeJson.java#L111-L115

alcatrazEscapee replied:
Well that's a nuisance. In which case, [here](https://github.com/DisasterMoo/TerraFirmaCraft/blob/ef7fb80e6e89f5c618d761e637c5712b78170d73/src/main/java/net/dries007/tfc/util/json/VeinTypeJson.java#L136), switch the ordering of that call chain to:
```java
if (ForgeRegistries.BLOCKS.containsKey(oreName)) {
    Block block = ForgeRegistries.BLOCKS.getValue(oreName); // Blocks have a nonnull default key, we can ignore null checks on this
    ...
```
as that should more accurately handle this case:
 - no arbitrary restrictions on the usage
 - accurate error reporting if something invalid was specified

alcatrazEscapee commented on src/main/java/net/dries007/tfc/world/classic/worldgen/vein/VeinType.java:
```diff
@@ -214,4 +227,61 @@ public IBlockState getOreState(Rock rock, Ore.Grade grade)
         }
     }
+
+    public enum Rarity
```
What is the point of this? Rarity can already be displayed as an integer, and you're enforcing names upon specific values that may have no applicability to whatever custom veins someone is using.

DisasterMoo replied:
Seems you lost the point when i was talking about it in discord. Plain number does nothing and is horrible to the average user, while names are actually helpful to them. Here are both screenshots when i was talking about it:
![screen1](https://user-images.githubusercontent.com/22617477/77234886-817ae900-6b90-11ea-8ae5-91adc9cf8558.png)
![screen2](https://user-images.githubusercontent.com/22617477/77234888-83dd4300-6b90-11ea-8474-00c3f03cd390.png)

And yes, I can just remove it and show nothing, but why not give more info to the player if possible?

alcatrazEscapee replied:
From discord:
> but what does "Rare" mean

This is exactly the point. You aren't showing more information - you're just reducing the resolution of the rarity. You could well display "Rarity: 3%", (taking `1 / rarity`) rather than a value, but "Rare", or any word you choose, is obscuring meaning and also, *completely dependent on a specific set of veins*.

alcatrazEscapee commented on src/main/java/net/dries007/tfc/util/json/VeinTypeJson.java:
```diff
@@ -68,6 +86,49 @@ public VeinType deserialize(JsonElement json, Type typeOfT, JsonDeserializationC
+                    else
+                    {
+                        looseRock = new ItemStack(looseBlock, 1, metaLoose);
+                    }
```
This whole sequence is unused, because you construct an `ItemStack` anyway, which will either:
a. convert the `Block` to it's respective `Item` (which, if it exists, would've been caught from the item query above), or
b. if the block doesn't exist, it will return an empty item stack.

DisasterMoo committed:
A few changes for PR:
- Added missing checks for vein registry.
- Refactored find veins to be less ugly.
- Find veins command now uses proper lang entries where possible.
- Removed veinsinfo command.

DisasterMoo committed:
More changes for PR:
- TFCGuiAreas are now a proper class (Renamed to TFCInventoryGuiHandler as requested).
- Chisel recipe wrapper now set output to the debug block if failed to pick block.
- Fixed vein registry parsing invalid ore blocks.

DisasterMoo committed:
Changed how to show rarity in JEI vein category.

DisasterMoo committed:
Fixed falling blocks not destroying loose items if they fall from less than 2 blocks (Closes #902).

alcatrazEscapee commented on src/main/java/net/dries007/tfc/command/CommandFindVeins.java:
```diff
@@ -99,7 +131,244 @@ public int getRequiredPermissionLevel()
+                    try
+                    {
+                        final File file = new File(fileName);
+                        final BufferedWriter writer = new BufferedWriter(new FileWriter(file));
```
```suggestion
                    try(BufferedWriter writer = new BufferedWriter(new FileWriter(file))
                    {
```

alcatrazEscapee commented on src/main/java/net/dries007/tfc/command/CommandFindVeins.java:
```diff
@@ -99,7 +131,244 @@ public int getRequiredPermissionLevel()
+                        }
+
+                        writer.close();
+
```
```suggestion
```

alcatrazEscapee commented on src/main/java/net/dries007/tfc/world/classic/worldgen/WorldGenOreVeins.java:
```diff
@@ -97,9 +99,19 @@ public void generate(Random random, int chunkX, int chunkZ, World world, IChunkG
+                    // Failed to generate, debug info
+                    // This can be by a number of factors, mainly because at each expected replacing position we didn't find a matching raw rock.
+                    // Some possible causes: Width / Height / Shape / Density / Y / Rock Layer
+                    TerraFirmaCraft.getLog().warn("Failed to generate vein '{}' in chunk ({}, {}). Vein center pos ({}x, {}y, {}z)", vein.getType().getRegistryName(), chunkX, chunkZ, vein.getPos().getX(), vein.getPos().getY(), vein.getPos().getZ());
```
```suggestion
                    TerraFirmaCraft.getLog().debug("Failed to generate vein '{}' in chunk ({}, {}). Vein center pos ({}x, {}y, {}z)", vein.getType().getRegistryName(), chunkX, chunkZ, vein.getPos().getX(), vein.getPos().getY(), vein.getPos().getZ());
```

alcatrazEscapee commented on src/main/java/net/dries007/tfc/world/classic/worldgen/vein/VeinRegistry.java:
```diff
@@ -141,6 +142,12 @@ public void postInit()
+                        if (ConfigTFC.GENERAL.debug)
+                        {
+                            TerraFirmaCraft.getLog().info("Registered new vein " + vein.toString());
+                        }
```
```suggestion
                            TerraFirmaCraft.getLog().info("Registered new vein " + vein.toString());
```

alcatrazEscapee commented on src/main/resources/assets/tfc/lang/en_us.lang:
```diff
@@ -104,6 +109,12 @@ tfc.tooltip.time_command_set_calendar_time=Set Calendar Time to %s.
 tfc.tooltip.skill_set=Set %s to level %.2f
 tfc.tooltip.skill_add=Added %.2f levels to %s
+tfc.tooltip.veins_command.dump_veins=Dumping all veins in specified radius, this is gonna take a while...
+tfc.tooltip.veins_command.rate_veins=Dumping vein rate in specified radius, this is gonna take a while...
```
```suggestion
tfc.tooltip.veins_command.rate_veins=Dumping vein rates in specified radius, this is gonna take a while...
```

alcatrazEscapee reviewed (changes requested):
A couple nitpicks

DisasterMoo committed:
More changes requested for PR.
